Implement tesh-timeout and tesh-long-running

domenkozar opened the pull request:
Fixes #24

zupo commented on README.md:
```diff
@@ -112,7 +112,9 @@ You can set a few other optional directives in the header line:
 - `tesh-platform`: specify on which platforms this session block should be tested (`linux`, `darwin`, `windows`),
-- `tesh-fixture`: a filename to save the current snippet.
+- `tesh-fixture`: a filename to save the current snippet,
+- `tesh-timeout`: number of second before a command timeouts (defaults to 30s),
```
```suggestion
- `tesh-timeout`: number of seconds before a command timeouts (defaults to 30s),
```

zupo commented on README.md:
```diff
@@ -112,7 +112,9 @@ You can set a few other optional directives in the header line:
-- `tesh-fixture`: a filename to save the current snippet.
+- `tesh-fixture`: a filename to save the current snippet,
+- `tesh-timeout`: number of second before a command timeouts (defaults to 30s),
+- `tesh-timeout-expected`: if set to yes, the commands are expected to timeout (useful for showing long running processes).
```
```suggestion
- `tesh-timeout-expected`: if set to `yes` or `true`, the commands are expected to timeout (useful for showing long running processes).
```

zupo commented on src/tesh/extract.py:
```diff
@@ -125,11 +127,16 @@ def extract(
+                        session.timeout = int(timeout)
+
+                    if timeout_expected := directives.get("timeout-expect"):
+                        session.timeout_expected = timeout_expected == "yes"
```
```suggestion
                        import distutils
                        session.timeout_expected = bool(distutils.util.strtobool(timeout_expected))
```

zupo commented on src/tesh/tests/fixtures/timeout.md:
````diff
@@ -1,6 +1,6 @@
 
-```shell-session tesh-session="foo"
+```shell-session tesh-session="foo" tesh-timeout="1"
 $ sleep 35
````
```suggestion
$ sleep 3
```

zupo commented:
A few minor suggestions posted as inline comments. Gimme +1 / -1 on them, and I'll polish them in, update tests, etc.

zupo commented:
I don't fully understand the usecase for `tesh-timeout-expected`. Couldn't I just set a very long timeout and effectively have the same outcome? Why do I need an extra directive?

domenkozar commented:
> I don't fully understand the usecase for `tesh-timeout-expected`. Couldn't I just set a very long timeout and effectively have the same outcome? Why do I need an extra directive?

If you set a long timeout, the command will eventually timeout and fail.

If you run a long-running process that never exists (for example in my use case, `devenv up`), you want to kill the program eventually.

I need to look if we can inspect the output after timeout happens and make sure that only the final command can timeout.

zupo commented:
Maybe `tesh-long-running` instead of `tesh-timeout-expected`?

zupo commented on src/tesh/extract.py:
```diff
@@ -125,11 +127,17 @@ def extract(
+                    if timeout := directives.get("timeout"):
+                        session.timeout = int(timeout)
+
+                    if timeout_expected := directives.get("timeout-expect"):
```
```suggestion
                    if timeout_expected := directives.get("timeout-expected"):
```

domenkozar commented:
Renamed to `tesh-long-running` and implemented logic to expect only the last command to timeout. Now the output is compared even when it timeouts.

Also added `--no-debug` flag to be able to manually test this one.

Tests still fail for some reason :(

zupo commented:
> Tests still fail for some reason :(

I'll try to take a look at the tests tonight.

zupo commented on src/tesh/test.py:
```diff
@@ -103,6 +81,45 @@ def test(filename: str, session: ShellSession, verbose: bool, debug: bool) -> No
+
+    if not fnmatch.fnmatch(actual_output, expected_output):
+        print("❌ Failed")  # noqa: ENC100
+        print()
```
Forgot to print failing command here.

zupo reviewed (changes requested):
I pushed addition of timeout and long-running examples to README. Further fixes are needed though:

1. fail output no longer contains the command that was executed (hence `test_fail` test fails)
2. something changed in `tesh-ps1` that makes the `prompt` example in README.md to fail, and `test_prompt` as well
3. by default, when `tesh-long-running` is set, I would set `tesh-timeout` to 1, to make the UX a bit nicer (not have to wait 30 seconds)

domenkozar commented:
> I pushed addition of timeout and long-running examples to README. Further fixes are needed though:
> 
> 1. fail output no longer contains the command that was executed (hence `test_fail` test fails)
> 2. something changed in `tesh-ps1` that makes the `prompt` example in README.md to fail, and `test_prompt` as well
> 3. by default, when `tesh-long-running` is set, I would set `tesh-timeout` to 1, to make the UX a bit nicer (not have to wait 30 seconds)

3. That would be surprising behavior, what if your command takes 10s to run? I think whoever is running those tests can set the timeout to lower if they get annoyed by the long running times.

zupo commented:
@domenkozar: what's the status on this one? How can I help getting it ready for merge?

domenkozar commented:
@blueracer should award me a superman hero badge here

zupo commented:
> @blueracer-io should award me a superman hero badge here

You don't get an award for removing the 30sec slowdown you introduced in a recent PR, lol :P

zupo commented:
Fail for me locally with:

```console
➜  tesh git:(88385da) tesh README.md 
📄 Checking README.md
Traceback (most recent call last):
  File "/nix/store/mvlayd2cp4lpgwfsmbdhva7ssa3h9vwn-python3.10-poetry2nix-env-scripts/bin/.tesh-wrapped", line 13, in <module>
    sys.exit(tesh())
  File "/nix/store/pwzxy20l2sqc9lh7abrdik6vbxpyrn6v-python3-3.10.4-env/lib/python3.10/site-packages/click/core.py", line 1130, in __call__
    return self.main(*args, **kwargs)
  File "/nix/store/pwzxy20l2sqc9lh7abrdik6vbxpyrn6v-python3-3.10.4-env/lib/python3.10/site-packages/click/core.py", line 1055, in main
    rv = self.invoke(ctx)
  File "/nix/store/pwzxy20l2sqc9lh7abrdik6vbxpyrn6v-python3-3.10.4-env/lib/python3.10/site-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/nix/store/pwzxy20l2sqc9lh7abrdik6vbxpyrn6v-python3-3.10.4-env/lib/python3.10/site-packages/click/core.py", line 760, in invoke
    return __callback(*args, **kwargs)
  File "/Users/zupo/work/tesh/src/tesh/__init__.py", line 46, in tesh
    sessions = extract(f)
  File "/Users/zupo/work/tesh/src/tesh/extract.py", line 80, in extract
    raise RuntimeError("Hit end-of-file prematurely. Syntax error?")
RuntimeError: Hit end-of-file prematurely. Syntax error?
```

zupo commented:
This is the culprit: https://github.com/OceanSprint/tesh/pull/30/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5L205

zupo commented:
@domenkozar: OK, got it to work in CI and my MacBook. Can you try if it work on your local pls?

domenkozar commented:
Works for me
